File: basic_dsa_two_hundred_sixty_nine.java
/**
 * Basic DSA Problems - Maximum Subarray
 * Find contiguous subarray with largest sum (Kadane's Algorithm)
 */
public class basic_dsa_two_hundred_sixty_nine {

    // Problem: Find contiguous subarray with maximum sum. Return that sum.
    // Example: [-2,1,-3,4,-1,2,1,-5,4] -> 6 (subarray [4,-1,2,1])

    /**
     * Kadane: at each i, either extend prev best or start new. maxEndHere = max(maxEndHere+x, x).
     * Time O(n), Space O(1)
     */
    public static int maxSubArray(int[] nums) {
        int maxHere = nums[0], maxSoFar = nums[0];
        for (int i = 1; i < nums.length; i++) {
            maxHere = Math.max(maxHere + nums[i], nums[i]);
            maxSoFar = Math.max(maxSoFar, maxHere);
        }
        return maxSoFar;
    }

    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("Problem: Maximum Subarray");
        System.out.println("=".repeat(60));

        System.out.println("[-2,1,-3,4,-1,2,1,-5,4] -> " + maxSubArray(new int[]{-2, 1, -3, 4, -1, 2, 1, -5, 4}));
        System.out.println("[1] -> " + maxSubArray(new int[]{1}));
        System.out.println("[-1] -> " + maxSubArray(new int[]{-1}));
    }
}

File: basic_dsa_two_hundred_sixty_nine.cpp
/**
 * Basic DSA - Binary Search (LeetCode 704)
 * Return index of target in sorted array, or -1.
 */
#include <iostream>
#include <vector>
#include <algorithm>

int search(std::vector<int>& nums, int target) {
    auto it = std::lower_bound(nums.begin(), nums.end(), target);
    if (it != nums.end() && *it == target) return it - nums.begin();
    return -1;
}

int main() {
    std::cout << std::string(60, '=') << "\n";
    std::cout << "Problem: Binary Search\n";
    std::cout << std::string(60, '=') << "\n";

    std::vector<int> v1 = {-1, 0, 3, 5, 9, 12};
    std::cout << "[-1,0,3,5,9,12], 9 -> " << search(v1, 9) << "\n";

    std::vector<int> v2 = {-1, 0, 3, 5, 9, 12};
    std::cout << "[-1,0,3,5,9,12], 2 -> " << search(v2, 2) << "\n";

    std::vector<int> v3 = {5};
    std::cout << "[5], 5 -> " << search(v3, 5) << "\n";
}
